src/utils/collections.zig

```diff
@@ -64,6 +64,188 @@ pub fn RecyclingList(
     };
 }
 
+pub fn SplitUnionList(TaggedUnion: type) type {
+    const Tag = @typeInfo(TaggedUnion).Union.tag_type.?;
+
+    return struct {
+        lists: sig.utils.types.EnumStruct(Tag, List),
+
+        const Self = @This();
+
+        pub const Index = struct {
+            tag: Tag,
+            index: usize,
+        };
+
+        fn List(tag: Tag) type {
+            return std.ArrayListUnmanaged(FieldType(tag));
+        }
+
+        fn FieldType(tag: Tag) type {
+            inline for (@typeInfo(TaggedUnion).Union.fields) |field| {
+                if (std.mem.eql(u8, field.name, @tagName(tag))) {
+                    return field.type;
+                }
+            }
+        }
+
+        pub fn init() Self {
+            var lists: sig.utils.types.EnumStruct(Tag, List) = undefined;
+            inline for (@typeInfo(Tag).Enum.fields) |f| {
+                @field(lists, f.name) = .{};
+            }
+            return .{ .lists = lists };
+        }
+
+        pub fn deinit(self: *Self, allocator: Allocator) void {
+            inline for (@typeInfo(Tag).Enum.fields) |f| {
+                @field(self.lists, f.name).deinit(allocator);
+            }
+        }
+
+        pub fn append(self: *Self, allocator: Allocator, item: TaggedUnion) Allocator.Error!Index {
+            switch (@as(Tag, item)) {
+                inline else => |tag| {
+                    const list = self.listMut(tag);
+                    const unwrapped_item = @field(item, @tagName(tag));
+                    try list.append(allocator, unwrapped_item);
+                    return .{
+                        .tag = tag,
+                        .index = list.items.len - 1,
+                    };
+                },
+            }
+        }
+
+        pub fn addOne(self: *Self, allocator: Allocator, tag: Tag) Allocator.Error!Entry {
+            switch (tag) {
+                inline else => |comptime_tag| {
+                    const list = self.listMut(comptime_tag);
+                    return .{
+                        .ptr = try list.addOne(allocator),
+                        .index = .{
+                            .tag = tag,
+                            .index = list.items.len - 1,
+                        },
+                    };
+                },
+            }
+        }
+
+        pub const Entry = struct {
```

Comment:
maybe move this to be above `getEntry`

Reply:
My intent here was to put it right after the first method that uses it, which is `addOne`. Not super opinionated about it though.